src/lib.rs

```diff
@@ -1025,6 +1045,37 @@ impl<'a, K: 'a + Eq + Hash, V: 'a, S: 'a + BuildHasher + Clone> Map<'a, K, V, S>
         }
     }
 
+    fn _get_or_insert(&'a self, key: K, value: impl FnOnce(&K) -> V) -> Ref<'a, K, V, S> {
+        let hash = self.hash_usize(&key);
+
+        let idx = self.determine_shard(hash);
+
+        let shard = unsafe { self._yield_read_shard(idx) };
+
+        if let Some((kptr, vptr)) = shard.get_key_value(&key) {
+            unsafe {
+                let kptr: *const K = kptr;
+                let vptr: *const V = vptr.get();
+                Ref::new(shard, kptr, vptr)
+            }
+        } else {
+            drop(shard);
+
+            unsafe {
+                let c: K = ptr::read(&key);
+                let mut shard = self._yield_write_shard(idx);
+
+                shard.insert(key, SharedValue::new(value(&c)));
```

Comment:
i think there could be a race here? when multiple threads are racing with `get_or_insert()` for the same _missing_ key. all threads could reach here, resulting replacing writes from other threads? That's because it's not atomic to drop the read shard above and yield write shard here?